drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c

```diff
@@ -0,0 +1,668 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ng_at86rf2xx
+ * @{
+ *
+ * @file
+ * @brief       Netdev adaption for the AT86RF2xx drivers
+ *
+ * @author      Thomas Eichinger <thomas.eichinger@fu-berlin.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include "net/ng_ieee802154.h"
+#include "net/ng_netbase.h"
+#include "ng_at86rf2xx.h"
+#include "ng_at86rf2xx_netdev.h"
+#include "ng_at86rf2xx_internal.h"
+#include "ng_at86rf2xx_registers.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/* TODO: generalize and move to ng_ieee802154 */
+static size_t _make_data_frame_hdr(ng_at86rf2xx_t *dev, uint8_t *buf,
+                                   ng_netif_hdr_t *hdr)
+{
+    int pos = 0;
+
+    /* we are building a data frame here */
+    buf[0] = NG_IEEE802154_FCF_TYPE_DATA;
+    buf[1] = 0x88;      /* use short src and dst addresses as starting point */
+
+    /* if AUTOACK is enabled, then we also expect ACKs for this packet */
+    if (dev->options & OPT_AUTOACK) {
+        buf[0] |= NG_IEEE802154_FCF_ACK_REQ;
+    }
+
+    /* fill in destination PAN ID */
+    pos = 3;
+    buf[pos++] = (uint8_t)((dev->pan) & 0xff);
+    buf[pos++] = (uint8_t)((dev->pan) >> 8);
+
+    /* fill in destination address */
+    if (hdr->flags &
+        (NG_NETIF_HDR_FLAGS_BROADCAST | NG_NETIF_HDR_FLAGS_MULTICAST)) {
+        buf[pos++] = 0xff;
+        buf[pos++] = 0xff;
+    }
+    else if (hdr->dst_l2addr_len == 2) {
+        uint8_t *dst_addr = ng_netif_hdr_get_dst_addr(hdr);
+        buf[pos++] = dst_addr[1];
+        buf[pos++] = dst_addr[0];
+    }
+    else if (hdr->dst_l2addr_len == 8) {
+        buf[1] |= 0x04;
+        uint8_t *dst_addr = ng_netif_hdr_get_dst_addr(hdr);
+        for (int i = 7;  i >= 0; i--) {
+            buf[pos++] = dst_addr[i];
+        }
+    }
+    else {
+        /* unsupported address length */
+        return 0;
+    }
+
+    /* fill in source PAN ID (if applicable */
+    if (dev->options & OPT_USE_SRC_PAN) {
+        buf[pos++] = (uint8_t)((dev->pan) & 0xff);
+        buf[pos++] = (uint8_t)((dev->pan) >> 8);
+    } else {
+        buf[0] |= NG_IEEE802154_FCF_PAN_COMP;
+    }
+
+    /* fill in source address */
+    if (dev->options & OPT_SRC_ADDR_LONG) {
+        buf[1] |= 0x40;
+        buf[pos++] = (uint8_t)((dev->addr_long) >> 56);
+        buf[pos++] = (uint8_t)((dev->addr_long) >> 48);
+        buf[pos++] = (uint8_t)((dev->addr_long) >> 40);
+        buf[pos++] = (uint8_t)((dev->addr_long) >> 32);
+        buf[pos++] = (uint8_t)((dev->addr_long) >> 24);
+        buf[pos++] = (uint8_t)((dev->addr_long) >> 16);
+        buf[pos++] = (uint8_t)((dev->addr_long) >> 8);
+        buf[pos++] = (uint8_t)((dev->addr_long) & 0xff);
+    }
```

Comment:
> Ah yeah, I misread the ">>". One more reason to change this. :)

Misread it for what?

> Wouldn't memcpy introduce a byte-ordering issue here?

You are right. To quote `ng_netconf.h`:

```C
    /**
     * @brief    get/set long address in host byte order
     *
     * Examples for this include the EUI-64 in IEEE 802.15.4
     */
    NETCONF_OPT_ADDRESS_LONG,
```

Though I'm unsure why that is, since it complicates a lot.